web/server/network_check.go

```diff
@@ -0,0 +1,326 @@
+			continue
+		}
+
+		response := &stun.Message{}
+		if err := stun.Decode(rawResponse, response); err != nil {
+			logger.Errorw("Error decoding STUN message", "error", err)
+			continue
+		}
+
+		switch c := response.Type.Class; c {
+		case stun.ClassSuccessResponse:
+			var bindResponseAddr stun.XORMappedAddress
+			if err := bindResponseAddr.GetFrom(response); err != nil {
+				logger.Errorw("Error extracting address from STUN message", "error", err)
+				continue
+			}
+
+			// Check for transaction ID mismatch.
+			if bindRequest.TransactionID != response.TransactionID {
+				logger.Errorf("Transaction ID mismatch (expected %s, got %s)",
+					hex.EncodeToString(bindRequest.TransactionID[:]),
+					hex.EncodeToString(response.TransactionID[:]),
+				)
+				continue
+			}
+
+			stunResponse.BindResponseAddr = bindResponseAddr.String()
+			stunResponse.TimeToBindResponse = time.Since(bindStart)
+		case stun.ClassErrorResponse, stun.ClassIndication, stun.ClassRequest:
+			logger.Errorw("Unexpected STUN response received", "response_type", c)
+		}
+	}
+
+	logSTUNResults(logger, stunResponses, "UDP")
+	return nil
+}
+
+// Tests NAT over TCP against STUN servers.
+func testTCP(ctx context.Context, logger logging.Logger) error {
+	// Create a dialer with a consistent port (randomly chosen) from
+	// which to dial over tcp.
+	dialer := &net.Dialer{
+		Timeout: 5 * time.Second,
+		LocalAddr: &net.TCPAddr{
+			IP: net.ParseIP("0.0.0.0"),
+		},
+	}
+
+	// Build a STUN binding request to be used against all STUN servers.
+	bindRequest, err := stun.Build([]stun.Setter{
+		stun.TransactionID,
+		stun.BindingRequest,
+	}...)
+	if err != nil {
+		return err
+	}
+	bindRequestRaw, err := bindRequest.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	var stunResponses []*STUNResponse
+	for _, stunServerURLToTest := range stunServerURLsToTestTCP {
+		if ctx.Err() != nil {
+			logger.Info("Machine shutdown detected; stopping TCP network tests")
+			return nil
+		}
+
+		logger := logger.WithFields("stun_server_url", stunServerURLToTest)
+
+		// Unlike with UDP, TCP needs a new `conn` for every STUN server test (all
+		// derived from the same dialer that uses the same local address.)
+		conn, err := dialer.Dial("tcp", stunServerURLToTest)
+		if err != nil {
+			logger.Error("Error dialing STUN server via tcp")
+			continue
+		}
+
+		// `net.Conn`s do not function with contexts (only deadlines.) If passed-in context
+		// expires (machine is likely shutting down,) _or_ tests finish, close the underlying
+		// `net.Conn` asynchronously to stop ongoing network checks.
+		testTCPDone := make(chan struct{})
+		defer func() {
+			testTCPDone <- struct{}{}
+		}()
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-testTCPDone:
+			}
+			conn.Close() //nolint:gosec,errcheck
+		}()
+
+		// Set a deadline for this interaction of 5 seconds in the future.
+		if err := conn.SetReadDeadline(time.Now().Add(5 * time.Second)); err != nil {
```

Comment:
a stylistic nit, but I got concerned when I didn't* see this around the `Read`. I don't moving this down would break anything?